// ods2json/xmlvars.js
/**
 * OpenOffice / LibreOffice predefined document structure constants
 *
 * @type Object
 */
exports.vars = {
    root                     : '$',
    body                     : 'office:body',
    chart                    : 'office:chart',
    calculated_value         : 'office:value',
    styles                   : 'office:styles',
    settingsBody             : 'office:settings',
    cell_content_type        : 'office:value-type',
    annotation               : 'office:annotation',
    spreadsheet              : 'office:spreadsheet',
    documentStyles           : 'office:document-styles',
    font_face_section        : 'office:font-face-decls',
    css_section              : 'office:automatic-styles',
    documentContent          : 'office:document-content',
    documentSettings         : 'office:document-settings',
    configName               : 'config:name',
    configItem               : 'config:config-item',
    configList               : 'config:config-item-set',
    configMap                : 'config:config-item-map-entry',
    namedMap                 : 'config:config-item-map-named',
    indexedConfigMap         : 'config:config-item-map-indexed',
    tableName                : 'table:name',
    table                    : 'table:table',
    shapes                   : 'table:shapes',
    formula                  : 'table:formula',
    tr                       : 'table:table-row',
    table_cell               : 'table:table-cell',
    table_class_name         : 'table:style-name',
    table_column             : 'table:table-column',
    rowspan                  : 'table:number-rows-spanned',
    tr_repeats               : 'table:number-rows-repeated',
    colspan                  : 'table:number-columns-spanned',
    td_repeats               : 'table:number-columns-repeated',
    cell_default_class       : 'table:default-cell-style-name',
    class                    : 'style:name',
    styleName                : 'style:family',
    css                      : 'style:style',
    font_face                : 'style:font-face',
    fontName                 : 'style:font-name',
    styleDefaults            : 'style:default-style',
    textRotation             : 'style:rotation-angle',
    numbersAfterDot          : 'style:decimal-places',
    textCss                  : 'style:text-properties',
    tableProps               : 'style:table-properties',
    parentClass              : 'style:parent-style-name',
    shapeCssProps            : 'style:graphic-properties',
    paragraphCss             : 'style:paragraph-properties',
    tableRowProps            : 'style:table-row-properties',
    tableCellProps           : 'style:table-cell-properties',
    tableColProps            : 'style:table-column-properties',
    shapeLeft                : 'svg:x',
    shapeTop                 : 'svg:y',
    shapeWidth               : 'svg:width',
    shapeHeight              : 'svg:height',
    shapeViewBox             : 'svg:viewBox',
    fontFamily               : 'svg:font-family',
    ahref                    : 'text:a',
    text_value               : 'text:p',
    textSpan                 : 'text:span',
    textClass                : 'text:style-name',
    block_title              : 'draw:name',
    shapesFrame              : 'draw:frame',
    shapesImage              : 'draw:image',
    mediaPlugin              : 'draw:plugin',
    shapesObject             : 'draw:object',
    z_index                  : 'draw:z-index',
    svgEquations             : 'draw:equation',
    shapeModifiers           : 'draw:modifiers',
    annotationClass          : 'draw:style-name',
    customShapes             : 'draw:custom-shape',
    shapePath                : 'draw:enhanced-path',
    annotationTextClass      : 'draw:text-style-name',
    shapesGeometry           : 'draw:enhanced-geometry',
    shapesObjectRange        : 'draw:notify-on-update-of-ranges',
    fontSize                 : 'fo:font-size',
    graphicFont              : 'fo:font-family',
    chartData                : 'chart:chart',
    plot                     : 'chart:plot-area',
    'fo:color'               : 'color',
    'svg:stroke-color'       : 'color',
    'style:column-width'     : 'width',
    'default-cell-style-name': 'class',
    'style:row-height'       : 'height',
    'fo:border'              : 'border',
    annotationDate           : 'dc:date',
    'style:family'           : 'display',
    'style:font-name'        : 'font-name',
    'fo:font-size'           : 'font-size',
    shapeImageURI            : 'xlink:href',
    'fo:font-style'          : 'font-style',
    'fo:text-align'          : 'text-align',
    'fo:border-top'          : 'border-top',
    'fo:font-weight'         : 'font-weight',
    'fo:border-left'         : 'border-left',
    'fo:border-right'        : 'border-right',
    'fo:break-before'        : 'break-before',
    'fo:border-bottom'       : 'border-bottom',
    'draw:fill-color'        : 'background-color',
    'fo:background-color'    : 'background-color',
    svmChartCoords           : 'chartooo:coordinate-region'
};
